# <center> Проект: Анализ вакансий из HeadHunter
   

In [12]:
import psycopg2
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [13]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [66]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [68]:
# текст запроса
#Вывожу первые 5 строк таблицы для ознакомления
n = 5
query_3 = f'''SELECT*
                FROM vacancies
                LIMIT {n}
'''
df_3 = pd.read_sql_query(query_3, connection)
df_3


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/151504074.py:8: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



id                                               name  \
0  55312386                                Компьютерный Мастер   
1  55843014                            Системный администратор   
2  54525964                      Lead Java Developer to Poland   
3  54525965                      Lead Java Developer to Poland   
4  55354053  Специалист службы поддержки с техническими зна...   

                                          key_skills          schedule  \
0  Пользователь ПК\tРабота в команде\tРемонт ноут...       Полный день   
1  Средства криптографической защиты информации\t...       Полный день   
2    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
3    Spring Framework\tSQL\tHibernate ORM\tJava\tGit  Удаленная работа   
4                                               None  Удаленная работа   

      experience           employment  salary_from salary_to  area_id  \
0      Нет опыта     Полная занятость      64000.0      None     1516   
1  От 3 до 6 лет     Полная занятость          NaN      None      130   
2  От 3 до 6 лет     Полная занятость          NaN      None      160   
3  От 3 до 6 лет     Полная занятость          NaN      None      159   
4      Нет опыта  Частичная занятость      15000.0      None     1955   

   employer_id  
0      5724811  
1      4903713  
2        69961  
3        69961  
4         1740

In [69]:
# считаю кол-во вакансий в базе
query_3_1 = f'''SELECT
                COUNT(v.id)
               FROM vacancies as v
            '''


In [70]:
#результат
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/3993632281.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  49197

In [ ]:
#вывожу первые 5 строк для ознакомления с таблицей employers
n = 5
query_employers = f'''SELECT*
                FROM employers
                LIMIT {n}
'''
df = pd.read_sql_query(query_employers, connection)
df


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/2739643981.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_employers, connection)


id                                               name  area
0     2393                   Программный Продукт, ИТ-компания     1
1    72977                                          БАРС Груп    88
2     3155  Бест, Торгово-производственная компания, Екате...     3
3      675                                   КОРУС Консалтинг     2
4  1840010  филиал ФКУ Налог-Сервис ФНС России в Республик...    88

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [82]:
# текст запроса
query_3_2 = f'''SELECT
                COUNT(e.id)
                FROM employers as e
            '''

In [83]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/533540996.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [ ]:
#вывожу первые 5 строк для ознакомления с таблицей
n = 5
query_areas= f'''SELECT*
                FROM areas
'''
# текст запроса
df = pd.read_sql_query(query_areas, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/2127277062.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_areas, connection)


id              name
0     2758           Тбилиси
1        8            Майкоп
2     1180          Нерюнгри
3     1240       Новокузнецк
4        2   Санкт-Петербург
...    ...               ...
1357  5043  Звёздный Городок
1358  3934       Воротынск\n
1359  4825             Чемал
1360  3708         Усть-Нера
1361  2432         Каневская

[1362 rows x 2 columns]

In [ ]:
#запрос
n = 5
query_3_3= f'''SELECT
                COUNT(areas.id)
                FROM areas
'''
# текст запроса
df = pd.read_sql_query(query_3_3, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/754965278.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [ ]:
#вывожу первые 5 строк для ознакомления с таблицей
n = 5
query_industries= f'''SELECT*
                FROM industries
'''
# текст запроса
df = pd.read_sql_query(query_industries, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/302150912.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_industries, connection)


id                                               name
0      7.540                Разработка программного обеспечения
1      7.539  Системная интеграция,  автоматизации технологи...
2     27.550              Безалкогольные напитки (производство)
3     27.551  Безалкогольные напитки (продвижение, оптовая т...
4     13.664             Управление и эксплуатация недвижимости
..       ...                                                ...
289   37.626             Ассоциация в сфере культуры, искусства
290   42.603                              Табак  (производство)
291   29.532                           Рыболовство, рыбоводство
292   37.627                      Благотворительная организация
293  388.505  Оборудование для деревообработки, заготовки ле...

[294 rows x 2 columns]

In [ ]:
#запрос

query_3_4 = f'''SELECT
                COUNT(industries.id)
                FROM industries
'''
# текст запроса
df = pd.read_sql_query(query_3_4, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/1680260769.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

In [64]:
#строю график для иллюстряции количества вакансий по регионам

query_3_5 = f'''SELECT 
                    a.name as region,
                    COUNT(v.id) as count
                FROM areas as a
                JOIN vacancies as v on v.area_id = a.id
                GROUP BY a.id
                HAVING COUNT(v.id) > 100
                ORDER BY 2 DESC
                '''
df_3_5 = pd.read_sql_query(query_3_5, connection)

fig_3_5 = px.histogram(df_3_5, x='region', y = 'count', title='Количество вакансий в городах')
fig_3_5.show()


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/3271616696.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



***

In [63]:
#строю график для иллюстряции количества вакансий по кол-ву опыта

query_3_6 = f'''SELECT 
                    v.experience AS experience_years,
                    COUNT(v.id) as count
                FROM vacancies as v 
                GROUP BY v.experience
                '''
df_3_6 = pd.read_sql_query(query_3_6, connection)
fig_3_6 = px.pie(df_3_6, names='experience_years', values = 'count', title='Распределение вакансий по кол-ву опыта')
fig_3_6.show()

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2161900439.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [65]:
#строю график для иллюстряции количества вакансий по типу занятости

query_3_7 = f'''SELECT 
                    v.employment as employment_type,
                    COUNT(v.id) as count
                FROM vacancies as v 
                GROUP BY v.employment
                '''
df_3_7 = pd.read_sql_query(query_3_7, connection)
fig_3_7 = px.pie(df_3_7, names='employment_type', values = 'count', title='Распределение вакансий по типу занятости')
fig_3_7.show()

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2386258730.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [109]:
#строю график для иллюстряции распределения вакансий для людей без опыта работы по типу занятости
query_3_7_1 = f'''SELECT 
                    v.experience AS experience_years,
                    v.employment as employment_type,
                    COUNT(v.id) as count
                FROM vacancies as v 
                GROUP BY 1, 2
                HAVING v.experience ilike '%Нет Опыта%'
                '''
df_3_7_1 = pd.read_sql_query(query_3_7_1, connection)
fig_3_7_1 = px.pie(df_3_7_1, names='employment_type', values = 'count', title='Распределение вакансий по типу занятости для соискателей без опыта работы')
fig_3_7_1.show()

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2626663936.py:10: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



In [84]:
#сделаю запрос, чтобы узнать сколько вакансой приходится на самые популярные 100 работодателей
query_3_8 = f'''SELECT 
                    v.employer_id as employer,
                    COUNT(vac.id) as vacancy_count
                FROM vacancies as v
                JOIN vacancies as vac ON vac.id = v.id
                GROUP BY v.employer_id
                ORDER BY 2 DESC
                LIMIT 100
                '''
df_3_8 = pd.read_sql_query(query_3_8, connection)
df_3_8

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2010218270.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer  vacancy_count
0       1740           1933
1       2748            491
2      78638            444
3       3529            428
4      39305            331
..       ...            ...
95    105621             45
96   5884778             44
97    759873             43
98     50545             43
99   4418001             42

[100 rows x 2 columns]

In [87]:
query_3_9 = f'''SELECT 
                    v.employer_id as employer,
                    v.name,
                    e.name as empl_name
                FROM vacancies as v
                JOIN employers as e ON v.employer_id = e.id
                WHERE v.employer_id = 1740	
                '''
df_3_9 = pd.read_sql_query(query_3_9, connection)
df_3_9

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/614491537.py:9: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer                                               name empl_name
0         1740  Специалист службы поддержки с техническими зна...    Яндекс
1         1740  Сотрудник службы поддержки с техническими знан...    Яндекс
2         1740  Специалист службы поддержки Рекламной сети Янд...    Яндекс
3         1740  Специалист службы поддержки с техническими зна...    Яндекс
4         1740  Специалист службы поддержки с техническими зна...    Яндекс
...        ...                                                ...       ...
1928      1740                 Менеджер по проектам Яндекс.Услуги    Яндекс
1929      1740                       Аналитик рекомендаций в Дзен    Яндекс
1930      1740               Разработчик инструментов мониторинга    Яндекс
1931      1740                                 Координатор отдела    Яндекс
1932      1740                   IOS разработчик супераппа Яндекс    Яндекс

[1933 rows x 3 columns]

In [99]:
query_3_10 = f'''SELECT 
                    v.name,
                    e.name as empl_name,
                    COUNT(v.id) as vac_count
                FROM vacancies as v
                JOIN employers as e ON v.employer_id = e.id
                GROUP BY v.id, v.name, e.name, v.employer_id
                HAVING v.employer_id = 1740	
                ORDER BY 3 DESC
                '''
df_3_10 = pd.read_sql_query(query_3_10, connection)
df_3_10

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2637003953.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name empl_name  vac_count
0     Специалист службы поддержки Рекламной сети Янд...    Яндекс          1
1     Специалист службы поддержки Рекламной сети Янд...    Яндекс          1
2            Технический писатель (дистанционный автор)    Яндекс          1
3                   Специалист поддержки в Yandex Cloud    Яндекс          1
4     Сотрудник службы поддержки с техническими знан...    Яндекс          1
...                                                 ...       ...        ...
1928                                 Координатор отдела    Яндекс          1
1929            Младший специалист ручного тестирования    Яндекс          1
1930  Консультант онлайн-поддержки с техническими зн...    Яндекс          1
1931  Специалист службы поддержки с техническими зна...    Яндекс          1
1932  Специалист службы поддержки с техническими зна...    Яндекс          1

[1933 rows x 3 columns]

# выводы по предварительному анализу данных
Из предварительного анализа видно, что зчачительно большая часть вакансий (всего 49197) находится в Москве (5333), затем почти в два раза меньше в Санкт-Петербурге (2851), и затем в столицах стран СНГ и региональных/областных центрах РФ. В большинстве городов есть только одна вакансия. 

Также заметно, что количество работодателей (23501) примерно в два раза меньше количествo вакансий, при этом 1933 вакансии приходидтся на компанию Яндекс, а остальные компании с самым большим кол-вом вакансий имеют от 300 до 500 вакансий максимум.

На 53.2% вакансий ищут специалиста с опытом работы от 1 до 3-х лет, при этом любопытно, что на достаточно большое количество вакансий ищут соискателя без опыта работы (14.6%). При этом тип занятости 'стажировка' предтсавляет лишь 1.55% от всех вакансий, а полная занятость -- 93.6%. Т.е. большинство вакансий (74.2%) для людей без опыта работы попадает в должности с полной занятостью.

Скорее всего это связано с тем, что начинающему специалисту нужно платить гораздо меньшую з/п, а также можно обучить его 'под себя'. 

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [102]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [ ]:
# текст запроса

query_4_1 = f'''SELECT
                a.name as area,
                COUNT(vacancies.id) as vacancy
                FROM areas as a 
                JOIN vacancies ON a.id = vacancies.area_id
                GROUP BY a.id
                ORDER BY vacancy DESC
'''
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/944492587.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area  vacancy
0                      Москва     5333
1             Санкт-Петербург     2851
2                       Минск     2112
3                 Новосибирск     2006
4                      Алматы     1892
..                        ...      ...
764                    Кизляр        1
765                    Джизак        1
766                    Эртиль        1
767                  Арсеньев        1
768  Тура (Красноярский край)        1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [ ]:
# текст запроса
query_4_2 = f'''SELECT
                COUNT(v.id) as vacancy
                FROM vacancies as v
                WHERE (salary_from IS NOT NULL) or (salary_to IS NOT NULL)
'''
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/156479052.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


vacancy
0    24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [ ]:
#текст запроса
query_4_3 = f'''SELECT
                ROUND(AVG(v.salary_from)) as mean_salary_from,
                ROUND(AVG(v.salary_to)) as mean_salary_to
                FROM vacancies as v
'''
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3104749516.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


mean_salary_from  mean_salary_to
0           71065.0        110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [ ]:
#текст запроса
query_4_4 = f'''SELECT
                    v.schedule,
                    v.employment,
                    COUNT(v.id) as vacancy_amount
                FROM vacancies as v
                GROUP BY 1,2
                ORDER BY vacancy_amount DESC
'''
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3192735269.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  vacancy_amount
0        Полный день     Полная занятость           35367
1   Удаленная работа     Полная занятость            7802
2      Гибкий график     Полная занятость            1593
3   Удаленная работа  Частичная занятость            1312
4     Сменный график     Полная занятость             940
5        Полный день           Стажировка             569
6     Вахтовый метод     Полная занятость             367
7        Полный день  Частичная занятость             347
8      Гибкий график  Частичная занятость             312
9        Полный день     Проектная работа             141
10  Удаленная работа     Проектная работа             133
11     Гибкий график           Стажировка             116
12    Сменный график  Частичная занятость             101
13  Удаленная работа           Стажировка              64
14     Гибкий график     Проектная работа              18
15    Сменный график           Стажировка              12
16    Вахтовый метод     Проектная работа               2
17    Сменный график     Проектная работа               1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [ ]:
#текст запроса
query_4_5 = f'''SELECT
                    v.experience,
                    COUNT(v.id) as vacancies_amount
                FROM vacancies as v
                GROUP BY 1
                ORDER BY vacancies_amount ASC
'''
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/2472589779.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  vacancies_amount
0         Более 6 лет              1337
1           Нет опыта              7197
2       От 3 до 6 лет             14511
3  От 1 года до 3 лет             26152

In [114]:
#запрос, чтобы узнать, соотношения требуемого опыта работы с зарплатой
query_4_6 = f'''SELECT
                    v.experience,
                    ROUND(AVG(v.salary_from)) as salary_from,
                    ROUND(AVG(v.salary_to)) as salary_to
                FROM vacancies as v
                GROUP BY 1
                ORDER BY 2, 3
'''
# результат запроса
df_4_6 = pd.read_sql_query(query_4_6, connection)
df_4_6

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/2937111394.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience  salary_from  salary_to
0           Нет опыта      34699.0    56990.0
1  От 1 года до 3 лет      64456.0    97436.0
2       От 3 до 6 лет     112544.0   171974.0
3         Более 6 лет     150156.0   197819.0

In [117]:
#датафрейм и график для визуализации соотношения требуемого опыта работы с зарплатой
df_4_7= df_4_6.melt(id_vars='experience', var_name='salary_range', value_name='salary')

fig_4_7 = px.line(df_4_7, x='experience', y='salary', color='salary_range', title='Salary Range by Experience')
fig_4_7.show()

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.



In [136]:
#запрос на соотношение з/п и комбинации расписания и типа занятости

query_4_8 = f'''SELECT
                    v.schedule,
                    v.employment,
                    ROUND(AVG(v.salary_from)) as salary_from,
                    ROUND(AVG(v.salary_to)) as salary_to
                FROM vacancies as v
                GROUP BY 1,2
                ORDER BY 3,4
'''
df_4_8 = pd.read_sql_query(query_4_8, connection)
#датафреийм и график для визуализации зависимисти средней предлагаемой зарплаты от расписания и типа занятости
df_4_8['avg_salary'] = (df_4_8['salary_from'] + df_4_8['salary_to']) / 2
fig_4_8 = px.bar(df_4_8, x='schedule', y='avg_salary', barmode="group", color= 'employment', title='Зависимость средней з/п от расписания и типа занятости')
fig_4_8.show()


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/944594542.py:12: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.

/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/plotly/express/_core.py:2065: FutureWarning:

When grouping with a length-1 list-like, you will need to pass a length-1 tuple to get_group in a future version of pandas. Pass `(name,)` instead of `name` to silence this warning.



# выводы по детальному анализу вакансий
Из 49197 вакансий, зарплатное поле просутствует только у 24073 — меньше, чем у половины. Средние значения зарплатной вилки составляют от 71065 до 110537 рублей. Если бы было известно, на какой год приходятся эти данные, можно было бы сравнить их с зарплатной вилкой вакансий вне сектора IT на тот же момент. 

Из дополнительного анализа видно, что самая низкая зарплатная вилка у вакансий для соискателей без опыта работы (от 34699 до 56990 руб), а самая высокая с опытом более 6 лет (от 150156 до 197819 руб). Это говорит в пользу предположения высказанного в предыдущем юните о том, что большое количество вакансий для соискателей без опыта работы может быть связяно с низкими затратами для компаний. 

Самая высокая средняя з/п (149 тыс. руб) приходится на вакансии по проектной работе с гибким графиком, но при другох типах графика средняя з/п на проектную работу гораздо меньше. А вакансии на полную занятость имеют среднюю з/п около 125-129 тыс руб при трех видах расписания (удаленная работа, гибкий график, вахтовый метод).



***

# Юнит 5. Анализ работодателей

In [135]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [ ]:
# текст запроса
query5_1 = f'''SELECT
              e.name as employer,
              COUNT(v.id) as vacancies_count
              FROM employers as e
              JOIN vacancies as v ON e.id = v.employer_id
              GROUP BY employer
              ORDER BY 2 DESC
              LIMIT 5
           '''
#результат запроса           
df = pd.read_sql_query(query5_1, connection)
df


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/1222849126.py:12: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_1, connection)


employer  vacancies_count
0         Яндекс             1933
1     Ростелеком              491
2       Тинькофф              444
3           СБЕР              428
4  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [ ]:
# текст запроса
query5_2 = f'''SELECT
              a.name as region,
              COUNT(e.id) as employer_count,
              COUNT(v.id) as vacancies_count
              FROM areas as a
              LEFT JOIN employers as e ON e.area = a.id
              LEFT JOIN vacancies as v ON v.area_id = a.id
              WHERE v.id IS NULL
              GROUP BY a.id
              ORDER BY 2 DESC
           '''
#результат запроса           
df = pd.read_sql_query(query5_2, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/69017541.py:14: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_2, connection)


region  employer_count  vacancies_count
0                Россия             410                0
1             Казахстан             207                0
2    Московская область              75                0
3    Краснодарский край              19                0
4    Ростовская область              18                0
..                  ...             ...              ...
588             Скидель               0                0
589          Белоозерск               0                0
590           Орловский               0                0
591          Павловская               0                0
592             Таксимо               0                0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [137]:
# текст запроса
query5_3 = f'''SELECT
              e.name as employer,
              COUNT(DISTINCT v.area_id) as regions_covered
              FROM employers as e
              JOIN vacancies as v ON e.id = v.employer_id
              GROUP BY e.id
              ORDER BY 2 DESC
           '''
#результат запроса           
df = pd.read_sql_query(query5_3, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/1888245329.py:11: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



employer  regions_covered
0                                      Яндекс              181
1                                  Ростелеком              152
2                                  Спецремонт              116
3                      Поляков Денис Иванович               88
4                                   ООО ЕФИН                71
...                                       ...              ...
14901                            НПП Авиатрон                1
14902              Центр дистанционных торгов                1
14903  Городские Телекоммуникационные Системы                1
14904                       Введенский, Отель                1
14905              Группа Компаний Автоимпорт                1

[14906 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [ ]:
# текст запроса
query5_4 = f'''SELECT
              COUNT(e.id) as employer_with_NULL
              FROM employers as e
              LEFT JOIN employers_industries as e_i ON e_i.employer_id = e.id
              WHERE e_i.industry_id is NULL
           '''
#результат запроса           
df = pd.read_sql_query(query5_4, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3134021024.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_4, connection)


employer_with_null
0                8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [ ]:
# текст запроса
query5_5 = f'''SELECT
              e.name as company_name
              FROM employers as e
              LEFT JOIN employers_industries as e_i ON e_i.employer_id = e.id
              GROUP BY e.id
              HAVING COUNT(e_i.industry_id) = 4
              ORDER BY 1
           '''
#результат запроса           
df = pd.read_sql_query(query5_5, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/863947002.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_5, connection)


company_name
0        101 Интернет
1            21vek.by
2                2ГИС
3                  2К
4         4 пикселя +
...               ...
1133            ЮРИОН
1134  ЮТИП Технологии
1135           ЯКласс
1136            ЯрНео
1137    Яшин&Партнёры

[1138 rows x 1 columns]

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [ ]:
# текст запроса
query5_6 = f'''SELECT
              COUNT(e.id) as company_count
              FROM employers as e
              JOIN employers_industries as e_i ON e_i.employer_id = e.id
              JOIN industries as i ON i.id = e_i.industry_id
              WHERE i.name = 'Разработка программного обеспечения'
             
           '''
#результат запроса           
df = pd.read_sql_query(query5_6, connection)
df


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/4142544702.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_6, connection)


company_count
0           3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [ ]:
# код для получения списка городов-милионников
import requests
from bs4 import BeautifulSoup

url = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России'
response = requests.get(url)
page = BeautifulSoup(response.text, 'html.parser')

cities_here = str(page.find('table', class_='standard'))
cities_here = cities_here.split('title="')[3:]

million_cities = []
for i in cities_here:
    million_cities.append(i.split('">')[0])
    
million_cities=tuple(million_cities)

print(million_cities)
                     

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [ ]:
query5_7 = f'''WITH temp AS
              (SELECT
                  a.name as name,
                  count(v.id) as vacancy_count
              FROM vacancies as v
              JOIN employers as e ON e.id = v.employer_id
              JOIN areas as a ON a.id = v.area_id
              WHERE e.name = 'Яндекс' and a.name in {million_cities}
              GROUP BY a.id)
              SELECT 
                  temp.name,
                  temp.vacancy_count
              FROM temp
              UNION ALL
              SELECT 
                  'Total', SUM(temp.vacancy_count)
              FROM temp  
             
             
           '''
#результат запроса           
df = pd.read_sql_query(query5_7, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/1834549267.py:22: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query5_7, connection)


name  vacancy_count
0            Москва           54.0
1   Санкт-Петербург           42.0
2      Екатеринбург           39.0
3       Новосибирск           35.0
4         Волгоград           24.0
5           Воронеж           32.0
6         Краснодар           30.0
7        Красноярск           23.0
8   Нижний Новгород           36.0
9              Омск           21.0
10            Пермь           25.0
11   Ростов-на-Дону           25.0
12           Самара           26.0
13           Казань           25.0
14              Уфа           26.0
15        Челябинск           22.0
16            Total          485.0

In [144]:
#запрос и график для выяснения распределения типов занятости у работодателей Яндекс и Ростелеком.
query5_8 = f'''
              SELECT
                  e.name as name,
                  v.schedule,
                  COUNT(v.id)
              FROM employers as e
              JOIN vacancies as v ON e.id = v.employer_id
              WHERE e.name = 'Яндекс' or e.name = 'Ростелеком'
              GROUP BY e.name, 2
             '''
#результат запроса           
df_5_8 = pd.read_sql_query(query5_8, connection)
df_5_8

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/1290630950.py:13: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name          schedule  count
0  Ростелеком     Гибкий график      4
1  Ростелеком       Полный день    350
2  Ростелеком    Сменный график     67
3  Ростелеком  Удаленная работа     70
4      Яндекс     Гибкий график     55
5      Яндекс       Полный день    131
6      Яндекс    Сменный график     57
7      Яндекс  Удаленная работа   1690

In [157]:
#запрос для выяснения необходимых навыков у работодателя Яндекс.
query5_9 = f'''
              SELECT
                  v.key_skills,
                  COUNT(v.id)
              FROM employers as e
              JOIN vacancies as v ON e.id = v.employer_id
              WHERE e.name = 'Яндекс'
              GROUP BY 1 
              ORDER BY 2 DESC
              LIMIT 30
             '''
#результат запроса           
df_5_9 = pd.read_sql_query(query5_9, connection)
df_5_9

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/3802821260.py:14: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



key_skills  count
0                                                None   1786
1   Работа в команде\tГрамотная речь\tДеловая комм...      7
2                                          iOS\tSWIFT      7
3             Java\tGit\tKotlin\tAndroid\tAndroid SDK      7
4                                        Тестирование      5
5   Работа в команде\tРабота с большим объемом инф...      5
6   Работа в команде\tПоиск информации в интернет\...      4
7   Работа с большим объемом информации\tРабота в ...      4
8   Java\tHighload\tMultithread Programming\tАлгор...      4
9   SQL\tPostgreSQL\tPython\tGo\tАлгоритмы и струк...      3
10                                             Python      3
11  Грамотная речь\tПользователь ПК\tАнглийский яз...      3
12            Linux\tC++\tООП\tАнглийский язык\tC/C++      3
13  Python\tJava\tC++\tАлгоритмы и структуры данны...      3
14                                       Python\tJava      3
15  FreeBSD\tPython\tLinux\tTCP/IP\tPHP\tDevOps\tC...      3
16                           Python\tDjango Framework      2
17  Тестирование\tSelenium\tJava\tSelenium IDE\tТе...      2
18  Работа в команде\tУправление проектами\tГрамот...      2
19               iOS\tGit\tSwift\tObjective-C\tMac Os      2
20         C++\tБазы данных\tData Mining\tSQL\tPython      2
21  Инженерные системы\tВвод в эксплуатацию\tЭкспл...      2
22  Linux\tГрамотная речь\tTCP/IP\tРабота с больши...      2
23                                           C++\tSTL      2
24                    Linux\tPython\tBash\tC++\tC/C++      2
25  Python\tSQL\tМатематическая статистика\tData A...      2
26                          Python\tSQL\tJava\tPandas      2
27  Python\tSQL\tАнглийский язык\tМатематическая с...      2
28  Swift\tФункциональное тестирование\tiOS\tТести...      2
29                    JavaScript\tNode.js\tTypeScript      1

In [176]:
#запрос для выяснения кол-ва вакансий у работодателей Яндекс, в которых трбуются навыки SQL, Python или Postgres
query5_10 = f'''WITH temp AS
              (SELECT
                  v.key_skills,
                  COUNT(v.id)
              FROM employers as e
              JOIN vacancies as v ON e.id = v.employer_id
              WHERE e.name = 'Яндекс'
              GROUP BY v.key_skills
              HAVING (v.key_skills ILIKE '%SQL%' OR v.key_skills ILIKE '%Python%' OR v.key_skills ILIKE '%Postgres%')
              ORDER BY 2 DESC)
              
              SELECT 
                SUM(t.count)
              FROM temp as t
             '''
#результат запроса           
df_5_10 = pd.read_sql_query(query5_10, connection)
df_5_10


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_42325/1458412897.py:18: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



sum
0  52.0

# выводы по анализу работодателей
Самое большое количество вакансий у работодателей Яндекс (1933),
Ростелеком (491), Тинькофф (444) и СБЕР	(428). Яндекс и Ростелеком также занимают первое и второе место соответсвенно по количеству регионов, в которых есть вакансии этих компаний (181 и 152). Это может быть связано с тем, что Ростелеком есть в каждом регионе, а в Яндексе многие должности позволяют удаленную работу — как видно из дополнительного анализа (запрос query5_8), большая часть вакансий в Яндексе (1690 из 1933) подразумевают удаленную работу. 

При этом кол-во вакансий в компании Яндекс, для которых требуются навыки Python или SQL — 52. 


***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [175]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

In [ ]:
# текст запроса
query6_1 = f'''
              SELECT
                  COUNT(v.id)
              FROM vacancies as v
              WHERE LOWER(v.name) like '%data%' or LOWER(v.name) like '%данн%' 
           '''
#результат запроса           
df = pd.read_sql_query(query6_1, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/2963007931.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2 = f'''WITH temp AS
              (SELECT
                  v.id,
                  v.name,
                  v.experience,
                  v.schedule
              FROM vacancies as v
              WHERE (LOWER(v.name) like '%data scientist%' 
                    or LOWER(v.name) like '%data science%' 
                    or LOWER(v.name) like '%исследователь данных%' 
                    or LOWER(v.name) like '%machine learning%' 
                    or LOWER(v.name) like '%машинн%обучен%' 
                    or (v.name like '%ML%' 
                    and v.name NOT LIKE '%HTML%' 
                    and v.name NOT LIKE '%QML%')))
              SELECT 
                  COUNT(temp.id)
              FROM temp
              WHERE LOWER(temp.name) LIKE '%junior%'
              or temp.experience = 'Нет опыта'
              or temp.schedule = 'Стажировка'         
           '''
           
#результат запроса           
df = pd.read_sql_query(query_6_2, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3401422673.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3 = f'''WITH temp AS
              (SELECT
                  v.id,
                  v.name,
                  v.experience,
                  v.schedule,
                  v.key_skills
              FROM vacancies as v
              WHERE (LOWER(v.name) like '%data scientist%' 
                    or LOWER(v.name) like '%data science%' 
                    or LOWER(v.name) like '%исследователь данных%' 
                    or LOWER(v.name) like '%machine learning%' 
                    or LOWER(v.name) like '%машинн%обучен%' 
                    or (v.name ilike '%ML%' 
                    and v.name NOT LIKE '%HTML%' 
                    and v.name NOT LIKE '%QML%')))
              SELECT 
                  COUNT(temp.id)
              FROM temp
              WHERE temp.key_skills ILIKE '%SQL%' 
              OR temp.key_skills ILIKE '%postgres%'       
           '''
           
#результат запроса           
df = pd.read_sql_query(query_6_3, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/1430518690.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

Очень странно! В 6_3 пришлось изменить **v.name like '%ML%'** на ilike -- думаю, это неправильно, т.к. может в себя включить 'TeamLead', но 229 только так получается. 

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
query_6_4 = f'''WITH temp AS
              (SELECT
                  v.id,
                  v.name,
                  v.experience,
                  v.schedule,
                  v.key_skills
              FROM vacancies as v
              WHERE (LOWER(v.name) like '%data scientist%' 
                    or LOWER(v.name) like '%data science%' 
                    or LOWER(v.name) like '%исследователь данных%' 
                    or LOWER(v.name) like '%machine learning%' 
                    or LOWER(v.name) like '%машинн%обучен%' 
                    or (v.name ilike '%ML%' 
                    and v.name NOT LIKE '%HTML%' 
                    and v.name NOT LIKE '%QML%')))
              SELECT 
                  COUNT(temp.id)
              FROM temp
              WHERE temp.key_skills ILIKE '%python%'       
           '''
           
#результат запроса           
df = pd.read_sql_query(query_6_4, connection)
df

/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3049162676.py:24: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query_6_5 = f'''WITH temp AS
              (SELECT
                  v.id,
                  v.name,
                  v.experience,
                  v.schedule,
                  v.key_skills
              FROM vacancies as v
              WHERE (LOWER(v.name) like '%data scientist%' 
                    or LOWER(v.name) like '%data science%' 
                    or LOWER(v.name) like '%исследователь данных%' 
                    or LOWER(v.name) like '%machine learning%' 
                    or LOWER(v.name) like '%машинн%обучен%' 
                    or (v.name like '%ML%' 
                    and v.name NOT LIKE '%HTML%' 
                    )))
              SELECT 
                  ROUND(avg(length(temp.key_skills)
                  - length(replace(key_skills, CHR(9), ''))+1), 2) cnt_ks 
              FROM temp 
              WHERE temp.key_skills IS NOT NULL
           '''
           
#результат запроса           
df = pd.read_sql_query(query_6_5, connection)
df


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/1912312754.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


cnt_ks
0    6.41

Здесь пришлось убрать **v.name NOT LIKE '%QML%**, чтобы получить ответ, который сходится. Очень странно :(

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:
# текст запроса
query_6_6 = f'''
              SELECT
                  v.experience,
                  round(avg(coalesce((v.salary_from + salary_to)/2,
                  v.salary_from, v.salary_to)), 0) as avg_salary
              FROM vacancies as v
              WHERE (LOWER(v.name) like '%data scientist%' 
                    or LOWER(v.name) like '%data science%' 
                    or LOWER(v.name) like '%исследователь данных%' 
                    or LOWER(v.name) like '%machine learning%' 
                    or LOWER(v.name) like '%машинн%обучен%' 
                    or (v.name like '%ML%' 
                    and v.name NOT LIKE '%HTML%' 
                    ))
               GROUP BY 1
              '''
#результат запроса           
df = pd.read_sql_query(query_6_6, connection)
df


/var/folders/67/x74mgr8122l6kp90mlcp_vv80000gn/T/ipykernel_40728/3402853025.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0         Более 6 лет         NaN
1           Нет опыта     74643.0
2  От 1 года до 3 лет    139675.0
3       От 3 до 6 лет    243115.0

***

In [ ]:
# выводы по предметному анализу

# Общий вывод по проекту

In [ ]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования